### PR TITLE
clientv3/naming: support resolving to multiple hosts

### PR DESCRIPTION
**Pull request by heyitsanthony**
Previous implementation watches a single key so there's no way to have separate hosts associate with separate keys for a single grpc target. Instead, accept all keys on a prefix.

Also fixes first the Next() to read current name data from etcd instead of waiting for the next event on a synced watcher.

### CODE REVIEW

File: Documentation/dev-guide/grpc_naming.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
+# gRPC naming and discovery
+
+etcd provides a gRPC resolver to support an alternative name system based that fetches endpoints from etcd for discovering grpc services. The underlying mechanism is based on watching updates to keys prefixed with the service name.
```
Comment (gyuho) [inline Documentation/dev-guide/grpc_naming.md]: Is that typo?

`s/system based that fetches/system that fetches/`

File: clientv3/naming/grpc.go [hunk 2]
```diff
@@ -16,99 +16,107 @@ package naming
...
 			updates = append(updates, &jupdate)
-		case mvccpb.DELETE:
-			updates = append(updates, &naming.Update{Op: naming.Delete})
 		}
 	}
+	return updates, nil
+}
+
+func (gw *gRPCWatcher) firstNext() ([]*naming.Update, error) {
+	resp, err := gw.c.Get(gw.ctx, gw.target, etcd.WithPrefix(), etcd.WithSerializable())
```
Comment (gyuho) [inline clientv3/naming/grpc.go]: Q. Why do we request with `WithSerializable`?

File: Documentation/dev-guide/grpc_naming.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+cli, cerr := clientv3.NewFromURL("http://localhost:2379")
+r := &etcdnaming.GRPCResolver{Client: cli}
+b := grpc.RoundRobin(r)
+conn, gerr := grpc.Dial("my-service", grpc.WithBalancer(b))
+```
+
+## Managing service endpoints
+
+The etcd resolver treats all keys under the prefix of the resolution target following a "/" (e.g., "my-service/") with JSON-encoded grpc `naming.Update` values as potential service endpoints. Endpoints are added to the service by creating new keys and removed frmo the service by removing keys.
```
Comment (xiang90) [inline Documentation/dev-guide/grpc_naming.md]: removed frmo -> removed form

File: clientv3/naming/grpc.go [hunk 4]
```diff
@@ -16,99 +16,107 @@ package naming
...
+	c      *etcd.Client
+	target string
+	ctx    context.Context
 	cancel context.CancelFunc
-	wch    clientv3.WatchChan
+	wch    etcd.WatchChan
+	err    error
 }
 
 func (gw *gRPCWatcher) Next() ([]*naming.Update, error) {
```
Comment (xiang90) [inline clientv3/naming/grpc.go]: what if people call next concurrently?

Comment (xiang90) [inline clientv3/naming/grpc.go]: do we need to protect gw.wch somehow?

**Comment by xiang90** [comment]
Fix the concurrently issue? Or simply doc it? Then look good to me.

File: clientv3/naming/grpc.go (hunk 2 above)
Comment (heyitsanthony) [inline clientv3/naming/grpc.go]: if the etcd server is servered from the quorum, anything using the resolver won't be able to grab a list of endpoints

Comment (gyuho) [inline clientv3/naming/grpc.go]: Ok makes sense. thanks.

File: clientv3/naming/grpc.go (hunk 4 above)
Comment (heyitsanthony) [inline clientv3/naming/grpc.go]: I can slap a lock on it, but I don't think calling it concurrently makes much sense; how would you reconcile an update sequence like add (a, 1.2.3.4) -> del a -> add (b, 1.2.3.4))?

Comment (heyitsanthony) [inline clientv3/naming/grpc.go]: right

**Pull request merged**